GATKReadFilterPluginDescriptor fixes

magicDGS opened the pull request:
- Disable default read filter with arguments does not longer blows up  but log a warning (fixes #2357)
- Improved help message for read filter (fixes #2358 and #2398)
- `--disableAllReadFilters` changed to `--disableToolDefaultReadFilters` and `isDisabledFilter()` honor this behavior (fixes #2361 and #2363)
- Add test for exercise the path of logging a warning when disabling a  not enabled filter (as proposed in #2377)
- Disabling a non-existent read filter throws a `CommandLineException`  (fixes #2397)

magicDGS commented:
This is the set of fixes for the filter plugin, @cmnbroad. The first commit is the change introduced in #2385 for less verbose test output, so it should be drop once it is accepted.

magicDGS commented:
I rebased into the latest changes to include the changes in #2385. Now this should be prepard for review, @cmnbroad! Thanks for everything!

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -221,6 +231,15 @@ public boolean isDependentArgumentAllowed(final Class<?> dependentClass) {
+            throw new CommandLineException.BadArgumentValue(
+                    String.format("The read filter(s) specified do not exists: ",
+                            Utils.join(", ", nonExistant)));
+        }
```
I'm assuming this is intended to be the fix for https://github.com/broadinstitute/gatk/issues/2397). I think it would be much cleaner and simpler to just integrate this check with the existing code below that already warns if a disabled filter is not enabled by the tool (I think you just need to add an if stmt there). Also, the error message (and comment, if you retain it) should specifically mention that its referring to a disabled filter.

magicDGS replied:
True, it is much cleaner than this implementation. I've done this to point out all the non-existent in the error, as a good way to remove in the same batch all possible invalid filters. Anyway, I simplified as you suggested and changed the error message.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -263,9 +282,11 @@ public void validateArguments() {
-            if (requiredPredecessors.contains(s)) {
+            if (requiredPredecessors.contains(s) && !toolDefaultReadFilters.containsKey(s)) {
                 throw new CommandLineException(
                         String.format("Values were supplied for (%s) that is also disabled", s));
```
There doesn't appear to be a test that executes this throw (I'm not sure if it was hit before, but since the if stmt is changed we should have tests that hit both clauses).

magicDGS replied:
There is a test case for throwing that is disabled because it fails after this change (I think that this is something related with Barclay, I added a `TODO` there for more information) and a new one covering the bug that I reported in #2357 (just logging). Thus, the only case where this could be exercised is when a non-default filter is disabled and arguments provided. Next commit will include that test.

cmnbroad replied:
Right, I update these code paths a bit and updated the tests for this so that there are test that execute all the code paths. The test you mentioned is still disabled, but I added comments about why we need to work around the Barclay behavior.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -293,11 +314,12 @@ public void validateArguments() {
     public boolean isDisabledFilter(final String filterName) {
-        return disableFilters.contains(filterName);
+        return disableFilters.contains(filterName)
+                || (disableToolDefaultReadFilters && !userReadFilterNames.contains(filterName));
```
I'm not sure why this change is needed  ? If it is, then the first line of the javadoc needs to change to reflect what it actually does.

magicDGS replied:
This is a fix for #2363. The new behavior of `--disableToolDefaultReadFilters` remove all the defaults filters only if the user one is not enable them again (if that change in `getAllInstances` is accepted). If the new behavior is removing all the default ones without allowing to re-enabled them, this line should be `(disableToolDefaultReadFilters && toolDefaultReadFilterNamesInOrder.contains(readFilter))`.

If someone is using this method to check if a required filter is disabled for a tool even if it is a default one, this should honor also the `--disableToolDefaultReadFilters`. Because the tool defaults were disabled in the command line if that parameter is provided, I think that it is still reflected in the first sentence. Anyway, I added in next commit a note saying that include disabled defaults if `--disableToolDefaultReadFilters` is provided in the command line.

cmnbroad replied:
Sounds good.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -348,13 +370,12 @@ public CountingReadFilter getMergedCountingReadFilter(final SAMFileHeader samHea
+                : toolDefaultReadFilterNamesInOrder
                 .stream()
                 .filter(s -> !isDisabledFilter(s))
                 .map(s -> toolDefaultReadFilters.get(s))
```
I think this was much clearer before (I realize it had to change to account for the new meaning of disableToolDefaultReadFilters). I guess its partly a style thing, but the extra parens and lack of indentation makes it hard to read; I think its much easier to parse like this:

```
        final List<ReadFilter> finalFilters = 
                disableToolDefaultReadFilters ?
                        new ArrayList<>(userReadFilterNames.size()) :
                        toolDefaultReadFilterNamesInOrder
                            .stream()
                            .filter(s -> !isDisabledFilter(s))
                            .map(s -> toolDefaultReadFilters.get(s))
                            .collect(Collectors.toList());
```

magicDGS replied:
Reformat done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -43,13 +43,17 @@ private void readObject(java.io.ObjectInputStream in)
+    // readFilters map is overridden in validateArguments
+    // this set is for keeping the names for the help in alphabetical order
+    // and to use them for checking if an argument exists or not
+    private final Set<String> readFilterNames = new TreeSet<>();
```
Do you mean to remove this set or just use a `HashSet` instead? There are two reasons for introducing this field:

* Previously, the help was different if command line validation was before or after the plugin validation. As I point out in my comments, the `readFilters` map is overriden in `validateArguments` for the ones that will be used. But the parsing system also perform custom validation of the tools after the validation of plugin arguments, so if something fails after that is called, the help will be incorrect. In addition, I did not explore other validation paths in Barclay, but it could happen that other validation is included after the plugin one and this may fail again to print all the filters.
* I also used this variable to check for existence of the plugin in the names, because relying in the `readFilters` name seems like a bad idea if the state is not maintained.

I will wait for the next round of comments for modifying this, just to clarify if I should remove it or change it for a hash-based set.

cmnbroad replied:
I meant the field should be removed altogether. readFilters is overwritten as you point out, but it doesn't need to be. So I've removed it in my set of changes.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -182,6 +187,9 @@ public boolean isDependentArgumentAllowed(final Class<?> dependentClass) {
                 filters.add(rf);
+            } else if (disableToolDefaultReadFilters) {
+                // include in the read instances since the tool one is disabled, but it is provided by the user
+                filters.add(toolDefaultReadFilters.get(s));
```
In this PR the meaning of disable "all" the filters changed to disable only the defaults. I tried to solve the following problems by this change:

* What happen if `--disableDefaultReadFilters --readFilter DEFAULT_FILTER_NAME` is specified? I think that the user provided filters must be applied independently of the default set.
* How can an user specify always the same parameters independently of any change in the tool default filters? I think that one of the points of the new behavior for this parameter is to provide a solution to that question. Disabling default filters one by one is not possible if they change over time, because that requires changes in the command line too; thus, this is a shortcut to disabling all the default read filters without knowing them, and provide always the same set of filters to apply. 
* How can an user reorganize default filters to apply? The order of filtering may be important in some case, for instance if one of them is computationally expensive with respect to the others. As an example: there is only one tool default filter, which is expensive and the user wants to pre-filter by a less demanding one. Previously it was not possible, but with this change the command line will provide the following: `--disableToolDefaultReadFilters --readFilter CUSTOM --readFilter DEFAULT_EXPENSIVE`

I think that this proposal is worthy from the user point of view, even if the new contract is not yet applied. Either that or this change should go to the `getMergedReadFilter()`to allow the behavior explained above.

It is true that there is no test case for this, I've just implemented one.

cmnbroad replied:
The use cases you mention are reasonable, but I think just making this change will result in duplicates in the merged filter. I think my updated changes handle this (I also added a test for the first bullet point).

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -366,7 +387,7 @@ public CountingReadFilter getMergedCountingReadFilter(final SAMFileHeader samHea
         }
 
-        return aggregateFunction.apply(finalFilters, samHeader);
+        return (finalFilters.isEmpty()) ? aggregateFunction.apply(null, samHeader) : aggregateFunction.apply(finalFilters, samHeader);
```
Good catch! It is something that I changed for whatever reason in one of my multiple PRs and branches fixing the plugin and kept here because I didn't realize...

Simplified in next commit.

magicDGS replied:
In addition, I added to the javadoc that the aggregateFunction should handle the empty list in the same way as `null`.

magicDGS commented:
Addressed most of the issues in the first pass. I have some questions/explanations for few of your comments, so back to you @cmnbroad. And thanks for discussing these issues!

lbergelson commented:
@magicDGS The cloud tests are skipping now for your builds so they shouldn't all be marked as failures incorrectly.  I was hoping that was fixed now.

magicDGS commented:
Awesome @lbergelson! Now the tests are passing here, so I will probably rebase all of my PRs soon to avoid the annoying "red cross of death". Thanks!

magicDGS commented:
What is the progress on this @cmnbroad? Is this waiting for the new Barclay plugin interface?

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -80,6 +76,9 @@ public GATKReadFilterPluginDescriptor(final List<ReadFilter> toolDefaultFilters)
                     className = rfClass.getName();
                 }
+                // we want to remember the order in which these were provided, and also keep the
+                // actual instances in case they have any additiona any state provided by the tool
```
"any additiona any state" -> "any additional state"

magicDGS replied:
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -127,70 +126,63 @@ public Object getInstance(final Class<?> pluggableClass) throws IllegalAccessExc
+     * order they were specified. Its possible for this to return a filter that was originally included
+     * in the list of tool defaults only in the case where the user also specifies it on the command line.
+     *
+     * NOTE: this method is somewhat misnamed in that it doesn't return ALL instances since it leaves out
```
We should add a note similar to this in `getDefaultInstances`

magicDGS replied:
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -231,62 +227,77 @@ public boolean isDependentArgumentAllowed(final Class<?> dependentClass) {
             }
         });
 
         // warn on redundant enabling of filters already enabled by default
```
This will read better like "warn if a filter is both default and enabled by the user"

magicDGS replied:
Done.

magicDGS commented on src/test/java/org/broadinstitute/hellbender/engine/filters/ReadFilterPluginUnitTest.java:
```diff
@@ -380,17 +384,15 @@ public void testDisabledDefaultWithArgsProvided() {
-        // get the command line read filters
-        clp.getPluginDescriptor(GATKReadFilterPluginDescriptor.class);
+        // Make sure ReadLengthReadFilter got disabled without an exception
+        Assert.assertTrue(rfDesc.userDisabledReadFilterNames.contains(ReadLengthReadFilter.class.getSimpleName()));
```
I will add also a test for `isDisabled`.

magicDGS replied:
Done.

magicDGS commented on src/test/java/org/broadinstitute/hellbender/engine/filters/ReadFilterPluginUnitTest.java:
```diff
@@ -259,20 +259,23 @@ public void testDisableOneFilter() {
                 "-disableReadFilter", ReadFilterLibrary.MAPPED.getClass().getSimpleName()
         });
+        // Make sure mapped filter got disabled with no exception
+        Assert.assertTrue(rfDesc.userDisabledReadFilterNames.contains(ReadFilterLibrary.MAPPED.getClass().getSimpleName()));
```
I will add also a test for isDisabled.

magicDGS replied:
Done.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -80,6 +76,9 @@ public GATKReadFilterPluginDescriptor(final List<ReadFilter> toolDefaultFilters)
+                // actual instances in case they have any additiona any state provided by the tool
+                // when they were created
                 toolDefaultReadFilterNamesInOrder.add(className);
                 toolDefaultReadFilters.put(className, f);
```
What do you think about using a `LinkedHashMap` instead of a list and a `HashMap`, @cmnbroad?

magicDGS commented on src/main/java/org/broadinstitute/hellbender/cmdline/GATKPlugin/GATKReadFilterPluginDescriptor.java:
```diff
@@ -36,24 +36,20 @@ private void readObject(java.io.ObjectInputStream in)
-    // this set is for keeping the names for the help in alphabetical order
-    // and to use them for checking if an argument exists or not
-    private final Set<String> readFilterNames = new TreeSet<>();
+    private Map<String, ReadFilter> allDiscoveredReadFilters = new HashMap<>();
```
What do you think if we make all these fileds "final" to avoid overriding as previously, @cmnbroad?

magicDGS commented on src/test/java/org/broadinstitute/hellbender/engine/filters/ReadFilterPluginUnitTest.java:
```diff
@@ -406,6 +408,22 @@ public void testDisabledNonDefaultFilterWithArgsProvided() {
     }
 
+    @Test
+    public void testDisableDefaultsAndReenableWithDifferentArgs() {
```
I will add also a test of disable all the defaults and re-enable with an user filter before it.

magicDGS replied:
Done.

magicDGS commented:
Back to you @cmnbroad. Your commit and some minor changes are included. There are still two questions:

* Why not using a `LinkedHashMap` instead of a `HashMap`/`ArrayList` for the default filters?
* Maybe it is better to make the fields final to do not override them in future changes or being aware of the change of state.

cmnbroad commented:
@magicDGS Both of those proposed changes (LinkedHashMap/final) sound right to me.

magicDGS commented:
Perfect, I include that changes in the latest commit. Back to you @cmnbroad.

magicDGS commented:
Thanks @cmnbroad!

cmnbroad commented:
@magicDGS Thanks for working through this!
